lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -622,20 +622,31 @@ def add_mpi_data_for_network_tokenization_card(post, payment, options)
 
         post[:mpiData] = {}
         post[:mpiData][:authenticationResponse] = 'Y'
-        post[:mpiData][:cavv] = payment.payment_cryptogram
+        if NETWORK_TOKENIZATION_CARD_SOURCE[payment.source.to_s].nil? && options[:switch_cryptogram_mapping_nt]
+          post[:mpiData][:tokenAuthenticationVerificationValue] = payment.payment_cryptogram
+        else
+          post[:mpiData][:cavv] = payment.payment_cryptogram
+        end
         post[:mpiData][:directoryResponse] = 'Y'
         post[:mpiData][:eci] = payment.eci || '07'
       end
 
-      def add_recurring_contract(post, options = {})
-        return unless options[:recurring_contract_type]
+      def add_recurring_contract(post, options = {}, payment = nil)
+        return unless options[:recurring_contract_type] || payment.is_a?(NetworkTokenizationCreditCard)
```

Comment:
I ran remote tests real quick on this for if `options[:switch_cryptogram_mapping_nt]=false` during an apple pay or google pay transaction. In that case we get an error from Adyen that says `"message"=>"Invalid recurring contract specified"` This is because the body has the following in it, and contract can't be nil if it is passed.  So we need to check for apple and google pay on the recurring contract logic as well.
```
{
  :recurring=>{ :contract=>nil }
}
```

```suggestion
        return unless options[:recurring_contract_type] || ( payment.is_a?(NetworkTokenizationCreditCard) && NETWORK_TOKENIZATION_CARD_SOURCE[payment&.source.to_s].nil? )
```

Reply:
After check the tests and verify the request contents I have tried a different approach because in my findings the fields `post[:recurring][:tokenService]` & `post[:recurring][:contract]` were not properly populated.
Please take a look and let me know what you think.